cuda/test/solver/lower_trs_kernels.cpp

```diff
@@ -0,0 +1,173 @@
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/solver/lower_trs.hpp>
+
+
+#include <memory>
+#include <random>
+
+
+#include <gtest/gtest.h>
+
+
+#include <cuda.h>
+
+
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/solver/lower_trs_kernels.hpp"
+#include "core/test/utils.hpp"
+
+
+namespace {
+
+
+class LowerTrs : public ::testing::Test {
+protected:
+    using CsrMtx = gko::matrix::Csr<double, gko::int32>;
+    using Mtx = gko::matrix::Dense<>;
+    LowerTrs() : rand_engine(30) {}
+
+    void SetUp()
+    {
+        ASSERT_GT(gko::CudaExecutor::get_num_devices(), 0);
+        ref = gko::ReferenceExecutor::create();
+        cuda = gko::CudaExecutor::create(0, ref);
+    }
+
+    void TearDown()
+    {
+        if (cuda != nullptr) {
+            ASSERT_NO_THROW(cuda->synchronize());
+        }
+    }
+
+    std::unique_ptr<Mtx> gen_mtx(int num_rows, int num_cols)
+    {
+        return gko::test::generate_random_matrix<Mtx>(
+            num_rows, num_cols,
+            std::uniform_int_distribution<>(num_cols, num_cols),
+            std::normal_distribution<>(-1.0, 1.0), rand_engine, ref);
+    }
+
+    std::unique_ptr<Mtx> gen_l_mtx(int num_rows, int num_cols)
+    {
+        return gko::test::generate_random_lower_triangular_matrix<Mtx>(
+            num_rows, num_cols, false,
+            std::uniform_int_distribution<>(num_cols, num_cols),
+            std::normal_distribution<>(-1.0, 1.0), rand_engine, ref);
+    }
+
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<const gko::CudaExecutor> cuda;
+    std::ranlux48 rand_engine;
+};
+
+
+TEST_F(LowerTrs, CudaLowerTrsFlagCheckIsCorrect)
+{
+    bool trans_flag = true;
+    bool expected_flag = false;
+#if (defined(CUDA_VERSION) && (CUDA_VERSION >= 9020))
+    expected_flag = false;
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+    expected_flag = true;
+#endif
+    gko::kernels::cuda::lower_trs::perform_transpose(cuda, trans_flag);
+
+    ASSERT_EQ(expected_flag, trans_flag);
+}
+
+
+TEST_F(LowerTrs, CudaSingleRhsApplyIsEquivalentToRef)
+{
+    std::shared_ptr<Mtx> mtx = gen_l_mtx(50, 50);
```

Comment:
You should move some of these definitions, at least the `mtx` as long as they can be shared into the `LowerTrs` test class.

Reply:
I would prefer if these would be separate as then you dont have any overlaps between the different tests. For the run-time, I think it would not hurt as the matrices are pretty small.

Reply:
They are always separate. [Gtest initiates the structure anew between every test](https://github.com/google/googletest/blob/master/googletest/docs/faq.md#should-i-use-the-constructordestructor-of-the-test-fixture-or-setupteardown), which is why we do this everywhere. For example, your own `omp/test/solver/lower_trs_kernels.cpp` use this model.

Reply:
But there it uses the `initialize_data()`. I could do the same here.

Reply:
The `initialize_data()` is here only when you need to control the size of the matrices. When you don't need to, everything could just go into the constructor or directly in the class definition, as, as the link from gtest says, they anyway always create a new test class for every test. The only time we would share anything would only be if we had any static member functions, I suppose.